### PR TITLE
code restructuring and adding basic xgboost model

### PR DESCRIPTION
**Pull request by navneet1075**
1. restructuring to allow more depth for multiple classification algorithms.
2. adding all the datasets in separate folder.

### CODE REVIEW

**Comment by kashyaprahul94** [comment]
I think you should rebase this branch on master first, so that `Updated readme` commit is not part of this PR, and you keep a linear commit history 😄

File: algorithms/.idea/workspace.xml [hunk 1]
```diff
@@ -1,119 +1,49 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (kashyaprahul94) [inline algorithms/.idea/workspace.xml]: I would take a stab and put `.idea` directory in `.gitignore` since it's very dev specific. What do you think ?

File: algorithms/classification/regression/preprocess.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 from sklearn.feature_selection import RFE
 
 def preprocess_train_data():
-    train_data = loaddata.load_train_data("./data/titanic-dataset","train.csv")
+    train_data = loaddata.load_train_data("../dataset/titanic","train.csv")
```
Comment (kashyaprahul94) [inline algorithms/classification/regression/preprocess.py]: This statement is repeated thrice in the file and requires you to enter file path again, how about moving this into a common method, say `get_data` ?

**Review by kashyaprahul94 — CHANGES_REQUESTED** [review]
Put some comments ☺️

File: algorithms/.idea/workspace.xml (hunk 1 above)
Comment (navneet1075) [reply algorithms/.idea/workspace.xml]: agreed . thats a very valid point . a very silly one . will correct this.

File: algorithms/classification/regression/preprocess.py (hunk 2 above)
Comment (navneet1075) [reply algorithms/classification/regression/preprocess.py]: good point. noted .

Comment (navneet1075) [reply algorithms/classification/regression/preprocess.py]: can be also done using a constant for the base path for dataset folder  and specific path for individual dataset can be provided as a parameter to the function and it can return the complete path .

Comment (kashyaprahul94) [reply algorithms/classification/regression/preprocess.py]: That's cherry on the cake 😉 I like it.

File: readme.md [hunk 3]
```diff
@@ -1,35 +1,38 @@
-ML Releated articles
```
Comment (kashyaprahul94) [inline readme.md:1]: think you should rebase this branch on master first, so that Updated readme commit is not part of this PR, and you keep a linear commit history 😄

**Comment by navneet1075** [comment]
deleting .idea was a pain literally, good lesson learned to properly use .gitignore and never ignore .gitignore 👽  .  merge to master will happen later after proper restructuring and documentation 🙈

**Review by kashyaprahul94 — COMMENTED** [review]
Looks good so far.

**Comment by kashyaprahul94** [comment]
> deleting .idea was a pain literally, good lesson learned to properly use .gitignore and never ignore .gitignore 👽 . merge to master will happen later after proper restructuring and documentation 🙈

Well, if i were to do it, i would have taken following approach -
* Always create a branch for unit of purpose. Let's say refactoring of something.
* A Pull Request should be `enough` to justify the purpose of the work as well as for the reviewer to have a good and solid understanding of the PR. Sometimes big changes tend to lead ignorance & rush 🤷‍♂ 
* Try to have smaller commits with meaningful commit messages because they are easy to follow. E.g. Added test for `get_data` function.
* At the end when about to merge, use the `squash` strategy which combines all the commits into one and then put it after the HEAD of master. That way you have a beautiful linear commit history 😍 

Considering your comment about `proper restructuring and documentation`, i would say there is no proper definition of proper 🤓 The idea of collaboration should be to continuous improvements in my personal view. That's said, you can always create a new PR for more restructuring and documentation in future, and prevent this PR from starvation 🗡 

Let me know if it makes sense :)

**Review by kashyaprahul94 — APPROVED** [review]

File: gini_impurity_and_gain.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+gini impurity:
```
Comment (kashyaprahul94) [inline gini_impurity_and_gain.md:1]: Please use Markdown syntax ❗️

File: decision_trees.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+overfitting in the decision trees :
```
Comment (kashyaprahul94) [inline decision_trees.md:1]: Please use Markdown syntax ❗️

**Comment by navneet1075** [comment]
- @kashyaprahul94 does make sense ,but whole intention here is to have complete things in master branch.that should be production grade code and as i far as i see, this repo is far from that stage . So just to add to your point , no big bang change will happen in the future. also no more content/feature will be added in this branch and only structure improvements will be done .

-  i would only work at this branch for now till the time a proper structure is not ready . 

-  things that was done wrong : 

-  i agree that i should have not done the restructuring in this branch and should have created a new branch for the change.  that will be followed going forward. 

- as far as the sqaush and merge goes , yes that is indeed a good approach and i will be doing it anyways for this branch.

Thanks for the valuable feedback 👍

**Review by kashyaprahul94 — COMMENTED** [review]
> * but whole intention here is to have complete things in master branch.that should be production grade code and as i far as i see, this repo is far from that stage .

Aah, that makes sense. I am happy then :)
